SPOI-6561 App Data Framework User Guide

davidyan74 opened the pull request.

sashadt commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,949 @@
+
+## Introduction
+
+The App Data Framework provides a way for data in Apex applications to be queried and delivered, so that the end user can easily access the data in the application and visualize it. 
```
"easily access the application data and visualize it".

sashadt commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,949 @@
+## Examples
+
+### A Simple Example
+In Apache Apex Malhar, the Twitter example demonstrates the usage of the App Data Framework. The application calculates top 10 hashtags mentioned in tweets in last 5 minutes across a 1% random tweet sampling on a rolling window basis.
```
"..in the last 5 minutes..".

sashadt commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,949 @@
+### A Simple Example
+In Apache Apex Malhar, the Twitter example demonstrates the usage of the App Data Framework. The application calculates top 10 hashtags mentioned in tweets in last 5 minutes across a 1% random tweet sampling on a rolling window basis.
+
+In DataTorrent RTS, you can create a dashboard from the application that has a bar chart widget with the current top hashtags in Twitter constantly updated:
```
"... create a dashboard for Twitter application ..."

tweise commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,954 @@
+
+Let’s look at the actual messages being exchanged for the Twitter example.
+
+#### SnapShot Schema
```
Snapshot

tweise commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,954 @@
+
+#### SnapShot Schema
+
+The SnapShot Schema servers a simple snapshot of tabular data, which is typically constantly being updated. This is the schema that is used in the Twitter example. We will describe the SnapShot schema as we look at the messages.
```
serves

tweise commented on docs/app_data_framework.md:
````diff
@@ -0,0 +1,954 @@
+}
+```
+
+The DataSource sends the dataResult message in response to the dataQuery. In this case, it contains a list of (hashtag, count) sets in the “data” field.
````
Datasource

davidyan74 commented:
All comments have been addressed. Please merge if you get a chance.

amberarrow commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,954 @@
+
+## Introduction
+
+The App Data Framework provides a way for data in Apex applications to be queried and delivered, so that the end user can easily access the application data and visualize it. 
```
Apache Apex ?

amberarrow commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,954 @@
+
+![Twitter Topology](images/app_data_framework/twitter_topology.png)
+
+The "SnapshotServer" operator (along with a Query Operator embedded in it) and the “QueryResult” operators enable the latest Twitter top hashtags to be visualizable in the widget. 
```
“QueryResult” operators => “QueryResult” operator

amberarrow commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,954 @@
+
+![Sales Topology](images/app_data_framework/sales_topology.png)
+
+Similar to the Twitter example, the “Store” operator along with the Query Operator embedded in it, and the “QueryResult” operators enable the data to be queried and be delivered to be visualized in the widget.
```
“QueryResult” operators => “QueryResult” operator

amberarrow commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,954 @@
+
+#### Data Source
+
+The Data Source Operator that runs in an Apex application that processes the queries from the Query Operator and gives back the results. In the Pi example, it is the “Snapshot Server” operator. The Java class of this operator in the Pi example is com.datatorrent.lib.appdata.snapshot.AppDataSnapshotServerMap in Apex Malhar. 
```
Apache Apex ?

davidyan74 replied:
I don't think we should say "Apache Apex" in place of Apex everywhere, especially when it already mentioned "Apache Apex" in the introduction... just like people don't usually say "Apache Spark" or "Apache Hadoop"

amberarrow commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,954 @@
+
+#### Query Operator
+
+The Query Operator runs embedded in the Data Source Operator, listens for incoming queries from the Message Bus and hands it over to the Data Source Operator. Note that since the Query Operator is embedded within the Data Source Operator, it will not be visible in the DAG view in dtConsole. In the Pi example, the Java class that implements the Query Operator is com.datatorrent.lib.io.PubSubWebSoc ...
```
Message Bus => Message Bus (see below)

amberarrow commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,954 @@
+
+#### Result Operator
+
+The Result Operator runs in an Apex application that publishes the results from the Data Source Operator to the Message Bus. In the Pi example, it is the “QueryResult” operator. The Java class of this operator in the Pi example is com.datatorrent.lib.io.PubSubWebSocketAppDataResult.
```
Apache Apex ?

Suggest using backquotes instead of double quotes since it highlights better

amberarrow commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,954 @@
+
+### Message Bus
+
+Because the operators can run in any nodes in the cluster, a Message Bus with a pub-sub mechanism is used for delivery for both the queries and the results. This mechanism requires that the caller sends the query message to a topic that the Query Operator listens to, and the Result Operator sends the query result message to a different topic that the caller listens to. Typically, the caller is a  ...
```
nodes => node

In the next sentence, it is not clear how the result makes it to the Result Operator.

amberarrow commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,954 @@
+
+Because the operators can run in any nodes in the cluster, a Message Bus with a pub-sub mechanism is used for delivery for both the queries and the results. This mechanism requires that the caller sends the query message to a topic that the Query Operator listens to, and the Result Operator sends the query result message to a different topic that the caller listens to. Typically, the caller is a  ...
+
+In DataTorrent RTS, this mechanism is provided by dtGateway. It uses WebSocket to achieve this so that the queries can be made and the results can be processed directly by a modern web browser. 
```
dtGateway => highlight with backquotes

amberarrow commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,954 @@
+
+### Schemas
+
+Each App Data Source must provide a schema that describes what is a valid data query and what is a valid data response, so that the caller knows how to query and how to interpret the response from the App Data Sources. We will describe two schemas that powers the Twitter example and the Sales example.
```
two schemas that powers the Twitter example and the Sales example. => two schemas that underlie the Twitter and Sales examples.
["powers" seems a bit too strong a word for this context]

amberarrow commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,954 @@
+
+The port that is returned by the embedded Query Operator’s `getOutputPort()` method will be connected to the input port of the Data Source Operator that is annotated with the marker annotation `com.datatorrent.common.experimental.AppData.QueryPort`.
+
+On the DAG level, the Result Operator needs to connect to an output port of the Data Source Operator that is annotated with the marker annotation com.datatorrent.common.experimental.AppData.ResultPort. 
```
com.datatorrent.common.experimental.AppData.ResultPort => highlight

amberarrow commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,954 @@
+
+With this setup, the Data Source, along with the query topic and the result topic, will be discoverable by STRAM. STRAM will return the information to the caller via its REST API.
+
+Below is the App Data Source discovery information returned by the `/ws/v2/applications/{appid}` REST call from dtGateway for the Twitter example:
```
dtGateway => highlight

amberarrow commented on docs/app_data_framework.md:
````diff
@@ -0,0 +1,954 @@
+}
+```
+
+The name field contains the name of the Data Source, and it is in the form of `<dataSourceOperatorName>.<resultPortName>`.
````
it is in the form of => has the following form:

amberarrow commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,954 @@
+
+“query” describes the query mechanism, including the topic. 
+
+“result” describes the result mechanism, including the topic and whether the query ID will be appended to the topic. The reason why appendQIDToTopic is true is because otherwise all callers would subscribe to the same result topic, and we do not want all callers to get the results for queries that they did not send. But for some message bus implementation, creating topics may be an expensive oper ...
```
appendIDToTopic => highlight

Suggest rewriting something like this:
Setting appendQIDToTopic to true increases the granularity of topics so that callers can
receive results only for their queries. If the cost of creating such a large number of topics
is high, it can be set to false; callers then will need to filter the results suitably.

amberarrow commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,954 @@
+
+### Message Bus
+
+dtGateway has a WebSocket service that provides the pubsub mechanism required by the Message Bus. The URL is ws://{dtGatewayAddress}/pubsub. You can publish data with a topic by sending a WebSocket message like this:
```
dtGateway => highlight
ws URL => ditto

amberarrow commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,954 @@
+4. The UI widget in the browser subscribes to the result topic
+5. The UI widget sends a data query
+6. The Data Source Operator gets the data query and processes it and sends back the data result to the result topic.
+7. The UI widgets receives the data result and renders it.
```
widgets => widget

amberarrow commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,954 @@
+
+#### Snapshot Schema
+
+The Snapshot Schema serves a simple snapshot of tabular data, which is typically constantly being updated. This is the schema that is used in the Twitter example. We will describe the Snapshot schema as we look at the messages.
```
updated. This is the schema that is used in the Twitter example. =>
updated and is used in the Twitter example.

amberarrow commented on docs/app_data_framework.md:
````diff
@@ -0,0 +1,954 @@
+```
+
+
+In preparation of the Schema Query, the browser first subscribes to the result topic (with the query ID appended) to get the schemaResult back. In this case, the browser generates a random string, “0.20716154835833223” for the query ID, and appends the query ID to the result topic that it is subscribing. 
````
In preparation of the => In preparation for a

schemaResult => Schema Result
[to be consistent with Schema Query]

that it is subscribing. => that it is subscribing to.

amberarrow commented on docs/app_data_framework.md:
````diff
@@ -0,0 +1,954 @@
+}
+```
+
+The browser sends a message with type, schemaQuery. This is the query that asks for the schema type and schema data. 
````
schemaQuery => highlight

amberarrow commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,954 @@
+
+The browser sends the dataQuery message that asks for actual data.
+
+The “countdown” field expects an optional integer value. It tells the Data Source that the Data Source should return results to them once for subsequent {countdown} application windows. In this example, the value is 30. That means the Data Source should execute this query 30 times, once for each application window for the next 30 application windows.
```
to them => omit

amberarrow commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,954 @@
+
+In this case, because the countdown value is 30 in the dataQuery, the Data Source will keep sending the latest data set for the next 29 application windows following this initial response.
+
+For more information about the operators behind this Snapshot Schema, please refer to this document.
```
Provide more precise reference.

amberarrow commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,954 @@
+This will query all the data with a “channel” value of “Mobile” and ANY value of “region”.
+Note that this is different from not specifying the “region” value at all, in which case, the aggregated value of all regions will be returned. This is similar to the difference between the SQL query of “SELECT SUM(sales) WHERE channel=’Mobile’” (for the case of not specifying “region”) and “SELECT SUM(sales), region WHERE channel=’Mobile’ GROUP BY region” (for the case of region: []).
+
+**fields**: This specifies what data to in the query result. If the field “time” is included, then a timestamp is included for each data point. Similarly the keys can also be specified as fields with the results. Lastly aggregated values can also be specified as fields to be returned in results.
```
data to in => data to include in

amberarrow commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,954 @@
+
+#### Dimension Schema: Additional Values for Combinations
+
+For the Dimension Schema supported by the DimensionsStore operator, there can be more complex schema specifications which have results for values only for certain combinations. This is the schema for yet another example about ads data aggregation in the DataTorrent RTS example repository.
```
ads data aggregation => aggregation of advertisement data

amberarrow commented on docs/app_data_framework.md:
````diff
@@ -0,0 +1,954 @@
+   "type":"data"
+}
+```
+Here you can see that a combination can also have additionalValues specified. Additional values are values which are only available for a specific combination of keys.
````
additionalValues => hightlight

davidyan74 commented:
@amberarrow Thank you so much for reviewing. I have addressed most of your comments. By the way, names like dtManage and dtGateway probably should not be backquoted since backquotes are usually used for things in the code like class names, function names, etc.

amberarrow commented:
There seems to be no uniform convention, so if you prefer to use the naked term, that's fine. I notice, for example that the Hadoop docs at https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-hdfs/HdfsUserGuide.html they use unadorned terms like Checkpoint, NameNode, DataNode.

Some guidelines here:
https://www.tu-chemnitz.de/phil/english/sections/linguist/independent/kursmaterialien/TechComm/acchtml/high.html

sashadt commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,956 @@
+
+![Twitter Topology](images/app_data_framework/twitter_topology.png)
+
+The *SnapshotServer* operator (along with a Query Operator embedded in it) and the *QueryResult* operator enable the latest Twitter top hashtags to be visualizable in the widget. 
```
Query Operator -> *Query* operator

sashadt commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,956 @@
+
+We will explain them in the following sections.
+
+The code of this Twitter example with App Data support is at [https://github.com/apache/incubator-apex-malhar/blob/master/demos/twitter/src/main/java/com/datatorrent/demos/twitter/TwitterTrendingHashtagsApplication.java](https://github.com/apache/incubator-apex-malhar/blob/master/demos/twitter/src/main/java/com/datatorrent/demos/twitter/TwitterTrendingHashtagsApplication.java)
```
is at [https://github.com/apache/incubator-apex-malhar/blob/master/demos/twitter/src/main/java/com/datatorrent/demos/twitter/TwitterTrendingHashtagsApplication.java] -> is available [here].

sashadt commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,956 @@
+
+### A More Complicated Example
+
+In DataTorrent RTS, the Sales example demonstrates a more complicated usage of the App Data Framework. The application generates random sales events, which has dimension keys of sales channels, region and products. Then it aggregates tax, sales amount and discount over the dimensions.
```
over the dimensions -> across all the key combinations.

sashadt commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,956 @@
+
+Because the operators can run in any node in the cluster, a Message Bus with a pub-sub mechanism is used for delivery for both the queries and the results. This mechanism requires that the caller sends the query message to a topic that the Query Operator listens to, and the Result Operator sends the query result message passed from the Data Source Operator to a different topic that the caller lis ...
+
+In DataTorrent RTS, this mechanism is provided by [dtGateway](http://docs.datatorrent.com/dtgateway/). It uses WebSocket to achieve this so that the queries can be made and the results can be processed directly by a modern web browser. 
```
modern web browser -> browser.  (No need to qualify this, since they can't work with RTS using an ancient web browser)

sashadt commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,956 @@
+
+In DataTorrent RTS, this mechanism is provided by [dtGateway](http://docs.datatorrent.com/dtgateway/). It uses WebSocket to achieve this so that the queries can be made and the results can be processed directly by a modern web browser. 
+
+Referring back to the architecture diagram, the web browser, the Query Operator and the Result Operator all connect to dtGateway via WebSocket, with a pub-sub protocol on top of it.
```
*Query* operator and the *Result* operator

sashadt commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,956 @@
+
+### UI Widgets
+
+In order to allow the end users to visualize the result data in a web browser, we need web-based UI widgets that take user input and send the queries and display the results according to the App Data Sources’ schemas. In DataTorrent RTS, we have a number of UI widgets built in. We will also describe how a user can write his/her own custom widgets.
```
his/her --> their

sashadt commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,956 @@
+
+### App Data Source
+
+As described in the overview, an App Data Source consists of three operators: Data Source Operator, Query Operator and the Result Operator.
```
three operators: Data Source, Query and the Result.

sashadt commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,956 @@
+
+In this case, because the countdown value is 30 in the *dataQuery*, the Data Source will keep sending the latest data set for the next 29 application windows following this initial response.
+
+For more information about the operators behind this Snapshot Schema, please refer to [this document](http://docs.datatorrent.com/operators/snapshot_server/).
```
Invalid link.  Is there a plan to write this doc soon?  If not, I suggest linking to Github code instead or deleting.

sashadt commented on docs/app_data_framework.md:
```diff
@@ -0,0 +1,956 @@
+
+For more information about the operators behind this Snapshot Schema, please refer to [this document](http://docs.datatorrent.com/operators/snapshot_server/).
+
+#### Dimension Schema
```
Name of the schema is Dimensions  (not Dimension)

sashadt commented:
Need to replace all subsequent occurrences of dimension --> dimensions when referring to schema.

sashadt commented on docs/app_data_framework.md:
````diff
@@ -0,0 +1,956 @@
+```
+This feature is supported by the DimensionsStore operator and is being used by the App Data Tracker.
+
+#### Dimension Schema: Additional Values for Combinations
````
Dimension -> Dimensions
